File: src/fr/cph/chicago/activity/MainActivity.java
/**
 * Copyright 2016 Carl-Philipp Harmant
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.activity;

import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.List;

import fr.cph.chicago.ChicagoTracker;
import fr.cph.chicago.R;
import fr.cph.chicago.data.BusData;
import fr.cph.chicago.data.DataHolder;
import fr.cph.chicago.entity.BikeStation;
import fr.cph.chicago.fragment.BikeFragment;
import fr.cph.chicago.fragment.BusFragment;
import fr.cph.chicago.fragment.FavoritesFragment;
import fr.cph.chicago.fragment.NearbyFragment;
import fr.cph.chicago.fragment.TrainFragment;
import fr.cph.chicago.task.LoadBusAndBikeDataTask;
import fr.cph.chicago.util.Util;

public class MainActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {

    private static final String SELECTED_ID = "SELECTED_ID";
    private static final int POSITION_BUS = 2;

    private int currentPosition;

    private Toolbar toolbar;
    private DrawerLayout drawerLayout;
    private ActionBarDrawerToggle drawerToggle;

    private FavoritesFragment favoritesFragment;
    private TrainFragment trainFragment;
    private BusFragment busFragment;
    private BikeFragment bikeFragment;
    private NearbyFragment nearbyFragment;

    private String title;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (!isFinishing()) {
            if (savedInstanceState != null) {
                ChicagoTracker.reloadData();
            }
            setContentView(R.layout.activity_main);

            new LoadBusAndBikeDataTask(this).execute();

            ChicagoTracker.container = (FrameLayout) findViewById(R.id.container);
            ChicagoTracker.container.getForeground().setAlpha(0);

            initView();
            setToolbar();

            drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
            drawerLayout.addDrawerListener(drawerToggle);
            drawerToggle.syncState();

            currentPosition = savedInstanceState == null ? R.id.navigation_favorites : savedInstanceState.getInt(SELECTED_ID);
            itemSelection(currentPosition);
        }
    }

    @Override
    public void onBackPressed() {
        if (currentPosition == R.id.navigation_favorites) {
            finish();
        } else {
            itemSelection(R.id.navigation_favorites);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        setBarTitle(this.title);
    }

    private void initView() {
        final NavigationView mDrawer = (NavigationView) findViewById(R.id.main_drawer);
        mDrawer.setNavigationItemSelectedListener(this);
        drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
    }

    private void setToolbar() {
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        toolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(final MenuItem item) {
                if (getString(R.string.nearby).equals(toolbar.getTitle())) {
                    nearbyFragment.reloadData();
                } else {
                    // Favorite fragment
                    if (favoritesFragment != null) {
                        favoritesFragment.startRefreshing();
                    }

                    Util.loadFavorites(favoritesFragment, FavoritesFragment.class, MainActivity.this);

                    // Google analytics
                    Util.trackAction(MainActivity.this, R.string.analytics_category_req, R.string.analytics_action_get_bus, R.string.analytics_action_get_bus_arrival, 0);
                    Util.trackAction(MainActivity.this, R.string.analytics_category_req, R.string.analytics_action_get_train, R.string.analytics_action_get_train_arrivals, 0);
                    Util.trackAction(MainActivity.this, R.string.analytics_category_req, R.string.analytics_action_get_divvy, R.string.analytics_action_get_divvy_all, 0);
                    // Check if bus/bike or alert data are not loaded. If not, load them.
                    // Can happen when the app has been loaded without any data connection
                    boolean loadData = false;
                    final DataHolder dataHolder = DataHolder.getInstance();

                    final BusData busData = dataHolder.getBusData();

                    final Bundle bundle = MainActivity.this.getIntent().getExtras();
                    final List<BikeStation> bikeStations = bundle.getParcelableArrayList(getString(R.string.bundle_bike_stations));

                    if (busData.getRoutes() != null && busData.getRoutes().size() == 0) {
                        loadData = true;
                    }
                    if (!loadData && bikeStations == null) {
                        loadData = true;
                    }
                    if (loadData) {
                        favoritesFragment.startRefreshing();
                        new LoadBusAndBikeDataTask(MainActivity.this).execute();
                    }
                    Util.trackAction(MainActivity.this, R.string.analytics_category_ui, R.string.analytics_action_press, R.string.analytics_action_refresh_fav, 0);
                }
                return true;
            }
        });
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            toolbar.setElevation(4);
        }
        toolbar.inflateMenu(R.menu.main);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().setNavigationBarColor(ContextCompat.getColor(this, R.color.primaryColorDarker));
        }
    }

    public void refresh(@NonNull final BusData busData, @NonNull final List<BikeStation> bikeStations) {
        // Put data into data holder
        final DataHolder dataHolder = DataHolder.getInstance();
        dataHolder.setBusData(busData);

        getIntent().putParcelableArrayListExtra(getString(R.string.bundle_bike_stations), (ArrayList<BikeStation>) bikeStations);
        onNewIntent(getIntent());
        if (favoritesFragment != null) {
            favoritesFragment.setBikeStations(bikeStations);
        }
        if (bikeFragment != null) {
            bikeFragment.setBikeStations(bikeStations);
        }
        if (currentPosition == POSITION_BUS && busFragment != null) {
            busFragment.update();
        }
    }

    private void setBarTitle(@NonNull final String title) {
        this.title = title;
        if (toolbar != null) {
            toolbar.setTitle(title);
        }
    }

    private void itemSelection(final int position) {
        final FragmentManager fragmentManager = getSupportFragmentManager();
        currentPosition = position;
        switch (position) {
            case R.id.navigation_favorites:
                setBarTitle(getString(R.string.favorites));
                if (favoritesFragment == null) {
                    favoritesFragment = FavoritesFragment.newInstance(position + 1);
                }
                if (!this.isFinishing()) {
                    fragmentManager.beginTransaction().replace(R.id.container, favoritesFragment).commit();
                }
                drawerLayout.closeDrawer(GravityCompat.START);
                showActionBarMenu();
                break;
            case R.id.navigation_train:
                setBarTitle(getString(R.string.train));
                if (trainFragment == null) {
                    trainFragment = TrainFragment.newInstance(position + 1);
                }
                if (!this.isFinishing()) {
                    fragmentManager.beginTransaction().replace(R.id.container, trainFragment).commit();
                }
                drawerLayout.closeDrawer(GravityCompat.START);
                hideActionBarMenu();
                break;
            case R.id.navigation_bus:
                setBarTitle(getString(R.string.bus));
                if (busFragment == null) {
                    busFragment = BusFragment.newInstance(position + 1);
                }
                if (!this.isFinishing()) {
                    fragmentManager.beginTransaction().replace(R.id.container, busFragment).commit();
                }
                drawerLayout.closeDrawer(GravityCompat.START);
                hideActionBarMenu();
                break;
            case R.id.navigation_bike:
                setBarTitle(getString(R.string.divvy));
                if (bikeFragment == null) {
                    bikeFragment = BikeFragment.newInstance(position + 1);
                }
                if (!this.isFinishing()) {
                    fragmentManager.beginTransaction().replace(R.id.container, bikeFragment).commit();
                }
                drawerLayout.closeDrawer(GravityCompat.START);
                hideActionBarMenu();
                break;
            case R.id.navigation_nearby:
                setBarTitle(getString(R.string.nearby));
                if (nearbyFragment == null) {
                    nearbyFragment = NearbyFragment.newInstance(position + 1);
                }
                if (!this.isFinishing()) {
                    fragmentManager.beginTransaction().replace(R.id.container, nearbyFragment).commit();
                }
                drawerLayout.closeDrawer(GravityCompat.START);
                showActionBarMenu();
                break;
        }
    }

    @Override
    public void onConfigurationChanged(final Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        drawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    public boolean onNavigationItemSelected(final MenuItem menuItem) {
        menuItem.setChecked(true);
        currentPosition = menuItem.getItemId();
        itemSelection(currentPosition);
        return true;
    }

    @Override
    public void onSaveInstanceState(final Bundle savedInstanceState) {
        super.onSaveInstanceState(savedInstanceState);
        savedInstanceState.putInt(SELECTED_ID, currentPosition);
        savedInstanceState.putString("title", title);
    }

    @Override
    public void onRestoreInstanceState(final Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        title = savedInstanceState.getString("title");
        currentPosition = savedInstanceState.getInt(SELECTED_ID);
    }

    private void hideActionBarMenu() {
        if (toolbar.getMenu().getItem(0).isVisible()) {
            showHideActionBarMenu(false);
        }
    }

    private void showActionBarMenu() {
        if (!toolbar.getMenu().getItem(0).isVisible()) {
            showHideActionBarMenu(true);
        }
    }

    private void showHideActionBarMenu(final boolean bool) {
        toolbar.getMenu().getItem(0).setVisible(bool);
    }
}

File: android-app/src/main/kotlin/fr/cph/chicago/core/activity/MainActivity.kt
/**
 * Copyright 2017 Carl-Philipp Harmant
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.core.activity

import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.os.Parcelable
import android.support.design.widget.NavigationView
import android.support.design.widget.Snackbar
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.MenuItem
import android.widget.FrameLayout
import butterknife.BindColor
import butterknife.BindString
import butterknife.BindView
import butterknife.ButterKnife
import fr.cph.chicago.Constants.Companion.BUSES_ARRIVAL_URL
import fr.cph.chicago.Constants.Companion.BUSES_ROUTES_URL
import fr.cph.chicago.Constants.Companion.TRAINS_ARRIVALS_URL
import fr.cph.chicago.R
import fr.cph.chicago.core.fragment.*
import fr.cph.chicago.entity.BikeStation
import fr.cph.chicago.rx.ObservableUtil
import fr.cph.chicago.service.BusService
import fr.cph.chicago.util.Util
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import java.util.*
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {

    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar
    @BindView(R.id.container)
    lateinit var frameLayout: FrameLayout
    @BindView(R.id.main_drawer)
    lateinit var drawer: NavigationView
    @BindView(R.id.drawer_layout)
    lateinit var drawerLayout: DrawerLayout

    @BindString(R.string.bundle_bike_stations)
    lateinit var bundleBikeStations: String
    @BindString(R.string.bundle_title)
    lateinit var bundleTitle: String
    @BindString(R.string.favorites)
    lateinit var favorites: String
    @BindString(R.string.train)
    lateinit var train: String
    @BindString(R.string.bus)
    lateinit var bus: String
    @BindString(R.string.divvy)
    lateinit var divvy: String
    @BindString(R.string.nearby)
    lateinit var nearby: String
    @BindString(R.string.cta_map)
    lateinit var ctaMap: String
    @BindString(R.string.cta_alert)
    lateinit var ctaAlert: String
    @BindString(R.string.settings)
    lateinit var settings: String

    @JvmField
    @BindColor(R.color.primaryColorDarker)
    internal var primaryColorDarker: Int = 0

    private val observableUtil: ObservableUtil = ObservableUtil
    private val util: Util = Util
    private val busService: BusService = BusService

    private var currentPosition: Int = 0

    private lateinit var drawerToggle: ActionBarDrawerToggle
    private lateinit var menuItem: MenuItem

    private var favoritesFragment: FavoritesFragment? = null
    private var trainFragment: TrainFragment? = null
    private var busFragment: BusFragment? = null
    private var bikeFragment: BikeFragment? = null
    private var nearbyFragment: NearbyFragment? = null
    private var ctaMapFragment: CtaMapFragment? = null
    private var alertFragment: AlertFragment? = null
    private var settingsFragment: SettingsFragment? = null

    private var title: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (!isFinishing) {
            if (savedInstanceState != null) {
                reloadData()
            }
            setContentView(R.layout.activity_main)
            ButterKnife.bind(this)

            loadFirstData()

            frameLayout.foreground.alpha = 0

            initView()
            setToolbar()

            drawerToggle = ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
            drawerLayout.addDrawerListener(drawerToggle)
            drawerToggle.syncState()

            currentPosition = savedInstanceState?.getInt(SELECTED_ID) ?: R.id.navigation_favorites
            itemSelection(currentPosition)

            checkForErrorInBundle()
        }
    }

    private fun checkForErrorInBundle() {
        val isTrainError = intent.getBooleanExtra(getString(R.string.bundle_train_error), false)
        val isBusError = intent.getBooleanExtra(getString(R.string.bundle_bus_error), false)
        // FIXME The snackbar does not move up the search button
        if (isTrainError && isBusError) {
            util.showSnackBar(this, R.string.message_something_went_wrong, Snackbar.LENGTH_SHORT)
        } else {
            if (isTrainError) {
                util.showSnackBar(this, R.string.message_error_train_favorites, Snackbar.LENGTH_SHORT)
            } else if (isBusError) {
                util.showSnackBar(this, R.string.message_error_bus_favorites, Snackbar.LENGTH_SHORT)
            }
        }
    }

    override fun onBackPressed() {
        if (currentPosition == R.id.navigation_favorites) {
            finish()
        } else {
            onNavigationItemSelected(menuItem)
        }
    }

    override fun onResume() {
        super.onResume()
        if(title != null)
            setBarTitle(title!!)
    }

    private fun initView() {
        drawer.setNavigationItemSelectedListener(this)
        menuItem = drawer.menu.getItem(0)
    }

    private fun setToolbar() {
        toolbar.setOnMenuItemClickListener { _ ->
            // Favorite fragment
            favoritesFragment!!.startRefreshing()

            util.trackAction(R.string.analytics_category_req, R.string.analytics_action_get_bus, BUSES_ARRIVAL_URL)
            util.trackAction(R.string.analytics_category_req, R.string.analytics_action_get_train, TRAINS_ARRIVALS_URL)
            util.trackAction(R.string.analytics_category_req, R.string.analytics_action_get_divvy, applicationContext.getString(R.string.analytics_action_get_divvy_all))
            util.trackAction(R.string.analytics_category_ui, R.string.analytics_action_press, applicationContext.getString(R.string.analytics_action_refresh_fav))

            if (util.isNetworkAvailable()) {
                if (busService.getBusRoutes().size == 0
                    || intent.getParcelableArrayListExtra<Parcelable>(bundleBikeStations) == null
                    || intent.getParcelableArrayListExtra<Parcelable>(bundleBikeStations).size == 0) {
                    loadFirstData()
                }
                val zipped = observableUtil.createAllDataObservable(application)
                zipped.subscribe({ favoritesResult -> favoritesFragment!!.reloadData(favoritesResult) })
                { onError ->
                    Log.e(TAG, onError.message, onError)
                    favoritesFragment!!.displayError(R.string.message_something_went_wrong)
                }
            } else {
                favoritesFragment!!.displayError(R.string.message_network_error)
            }
            true
        }
        toolbar.inflateMenu(R.menu.main)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            toolbar.elevation = 4f
            window.navigationBarColor = primaryColorDarker
        }
    }

    fun loadFirstData() {
        observableUtil.createOnFirstLoadObservable().subscribe(
            { (busRoutesError, bikeStationsError, busRoutes, bikeStations) ->
                busService.setBusRoutes(busRoutes)
                refreshFirstLoadData(bikeStations)
                if (bikeStationsError || busRoutesError) {
                    util.showSnackBar(this, R.string.message_something_went_wrong, Snackbar.LENGTH_SHORT)
                }
            },
            { _ -> util.showSnackBar(this, R.string.message_something_went_wrong, Snackbar.LENGTH_SHORT) }
        ) {
            util.trackAction(R.string.analytics_category_req, R.string.analytics_action_get_bus, BUSES_ROUTES_URL)
            util.trackAction(R.string.analytics_category_req, R.string.analytics_action_get_divvy, applicationContext.getString(R.string.analytics_action_get_divvy_all))
        }
    }

    private fun refreshFirstLoadData(bikeStations: List<BikeStation>) {
        intent.putParcelableArrayListExtra(bundleBikeStations, bikeStations as ArrayList<BikeStation>)
        onNewIntent(intent)
        if (favoritesFragment != null) {
            favoritesFragment!!.setBikeStations(bikeStations)
        }
        if (bikeFragment != null) {
            bikeFragment!!.setBikeStations(bikeStations)
        }
        if (currentPosition == POSITION_BUS && busFragment != null) {
            busFragment!!.update()
        }
    }

    private fun reloadData() {
        // Might want to reload train csv data if needed here
    }

    private fun setBarTitle(title: String) {
        this.title = title
        toolbar.title = title
    }

    private fun itemSelection(position: Int) {
        val fragmentManager = supportFragmentManager
        currentPosition = position
        when (position) {
            R.id.navigation_favorites -> {
                setBarTitle(favorites)
                favoritesFragment = if (favoritesFragment == null) FavoritesFragment.newInstance(position + 1) else favoritesFragment
                fragmentManager.beginTransaction().replace(R.id.container, favoritesFragment).commit()
                closeDrawerAndUpdateActionBar(true)
            }
            R.id.navigation_train -> {
                setBarTitle(train)
                trainFragment = if (trainFragment == null) TrainFragment.newInstance(position + 1) else trainFragment
                fragmentManager.beginTransaction().replace(R.id.container, trainFragment).commit()
                closeDrawerAndUpdateActionBar(false)
            }
            R.id.navigation_bus -> {
                setBarTitle(bus)
                busFragment = if (busFragment == null) BusFragment.newInstance(position + 1) else busFragment
                fragmentManager.beginTransaction().replace(R.id.container, busFragment).commit()
                closeDrawerAndUpdateActionBar(false)
            }
            R.id.navigation_bike -> {
                setBarTitle(divvy)
                bikeFragment = if (bikeFragment == null) BikeFragment.newInstance(position + 1) else bikeFragment
                fragmentManager.beginTransaction().replace(R.id.container, bikeFragment).commit()
                closeDrawerAndUpdateActionBar(false)
            }
            R.id.navigation_nearby -> {
                setBarTitle(nearby)
                nearbyFragment = if (nearbyFragment == null) NearbyFragment.newInstance(position + 1) else nearbyFragment
                Observable.create<Any> { subscriber ->
                    drawerLayout.closeDrawer(GravityCompat.START)
                    subscriber.onNext(Any())
                    subscriber.onComplete()
                }
                    .delay(500, TimeUnit.MILLISECONDS)
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnError { throwable -> Log.e(TAG, throwable.message, throwable) }
                    .subscribe { _ -> fragmentManager.beginTransaction().replace(R.id.container, nearbyFragment).commitAllowingStateLoss() }
                drawerLayout.closeDrawer(GravityCompat.START)
                hideActionBarMenu()
            }
            R.id.navigation_cta_map -> {
                setBarTitle(ctaMap)
                ctaMapFragment = if (ctaMapFragment == null) CtaMapFragment.newInstance(position + 1) else ctaMapFragment
                fragmentManager.beginTransaction().replace(R.id.container, ctaMapFragment).commit()
                closeDrawerAndUpdateActionBar(false)
            }
            R.id.alert_cta -> {
                setBarTitle(ctaAlert)
                alertFragment = if (alertFragment == null) AlertFragment.newInstance(position + 1) else alertFragment
                fragmentManager.beginTransaction().replace(R.id.container, alertFragment).commit()
                closeDrawerAndUpdateActionBar(false)
            }
            R.id.rate_this_app -> util.rateThisApp(this)
            R.id.settings -> {
                setBarTitle(settings)
                settingsFragment = if (settingsFragment == null) SettingsFragment.newInstance(position + 1) else settingsFragment
                fragmentManager.beginTransaction().replace(R.id.container, settingsFragment).commit()
                closeDrawerAndUpdateActionBar(false)
            }
        }
    }

    private fun closeDrawerAndUpdateActionBar(showActionBarMenu: Boolean) {
        drawerLayout.closeDrawer(GravityCompat.START)
        if (showActionBarMenu)
            showActionBarMenu()
        else
            hideActionBarMenu()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        drawerToggle.onConfigurationChanged(newConfig)
    }

    override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
        menuItem.isChecked = true
        currentPosition = menuItem.itemId
        if (!isFinishing)
            itemSelection(currentPosition)
        return true
    }

    public override fun onSaveInstanceState(savedInstanceState: Bundle?) {
        savedInstanceState!!.putInt(SELECTED_ID, currentPosition)
        savedInstanceState.putString(bundleTitle, title)
        super.onSaveInstanceState(savedInstanceState)
    }

    public override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        title = savedInstanceState.getString(bundleTitle)
        currentPosition = savedInstanceState.getInt(SELECTED_ID)
    }

    private fun hideActionBarMenu() {
        if (toolbar.menu.getItem(0).isVisible) {
            showHideActionBarMenu(false)
        }
    }

    private fun showActionBarMenu() {
        if (!toolbar.menu.getItem(0).isVisible) {
            showHideActionBarMenu(true)
        }
    }

    private fun showHideActionBarMenu(bool: Boolean) {
        toolbar.menu.getItem(0).isVisible = bool
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
        private const val SELECTED_ID = "SELECTED_ID"
        private const val POSITION_BUS = 2
    }
}